### PR TITLE
[WIP] Improve API docs

### PR DESCRIPTION
**Pull request by hawkw**
I've started working on some changes (hopefully, improvements!) to `tokio`'s API docs. These changes are still a work in progress, but I'm opening a draft PR to start getting feedback now.

### CODE REVIEW

File: tokio/src/task/mod.rs [hunk 1]
```diff
@@ -1,5 +1,202 @@
 //! Asynchronous green-threads.
-
+//!
+//! ## What are Tasks?
+//!
+//! A _task_ is a light weight, non-blocking unit of execution. A task is similar
+//! to an OS thread, but rather than being managed by the OS scheduler, they are
+//! managed by the [Tokio runtime][rt]. Key points about tasks include:
```
Comment (davidbarsky) [inline tokio/src/task/mod.rs]: This section is great! I'd note somewhere that Tokio's tasks are similar to goroutines. Something along the lines of:

- If you are familiar with [Go's goroutines](https://tour.golang.org/concurrency/1), [Kotlin's coroutines](https://kotlinlang.org/docs/reference/coroutines-overview.html), or [Erlang's processes](http://erlang.org/doc/getting_started/conc_prog.html#processes), you can think of Tokio's tasks as something similar.

File: tokio/src/task/mod.rs [hunk 2]
```diff
@@ -1,5 +1,202 @@
...
+//!   overhead. Creating, running, and destroying large numbers of tasks is
+//!   quite cheap, especially compared to OS threads.
+//!
+//! * Tasks are scheduled **cooperatively**. Most operating systems implement
+//!   _preemptive multitasking_. This is a scheduling technique where the
+//!   operating system allows each thread to run for a period of time, and then
+//!   _preempts_ it, temporarily pausing that thread and switching to another.
+//!   Tasks, on the other hand, implement _cooperative multitasking_. In
+//!   cooperative multitasking, a task is allowed to run until it _yields_,
+//!   indicating to the Tokio runtime's scheduler that it cannopt currently
```
Comment (davidbarsky) [inline tokio/src/task/mod.rs]:
```suggestion
//!   indicating to the Tokio runtime's scheduler that it cannot currently
```

File: tokio/src/task/mod.rs [hunk 3]
```diff
@@ -1,5 +1,202 @@
...
+//! [`task::spawn`]: crate::task::spawn()
+//! [future]: std::future::Future
+//! [`std::thread::spawn`]: std::thread::spawn
+//! [`JoinHandle`]: crate::task::JoinHandle
+//! [thread_join]: std::thread::JoinHandle
+//! [`JoinError`]: crate::task::JoinError
+//!
+//! ### Blocking and Yielding
+//!
+//! As we discussed above, code running in asynchronous tasks should nbt perform
```
Comment (davidbarsky) [inline tokio/src/task/mod.rs]:
```suggestion
//! As we discussed above, code running in asynchronous tasks should not perform
```

File: tokio/src/task/mod.rs (hunk 1 above)
Comment (hawkw) [reply tokio/src/task/mod.rs]: Done! https://github.com/tokio-rs/tokio/pull/1801/commits/467d4dcda961e882ac7c8434bdfc59620c048b40 Thanks for the suggestion!

File: tokio/src/task/mod.rs (hunk 2 above)
Comment (hawkw) [reply tokio/src/task/mod.rs]: thanks for catching these! fixed in 1873017b0aaf5c2b55524b36822c5f68e0158642

**Comment by hawkw** [comment]
Closing this, as I've made these additions in separate PRs (#1801, #1803, #1804, and #1806).

**Pull request closed without merging**